api/src/main/java/jakarta/persistence/Graph.java

```diff
@@ -32,28 +32,85 @@
 public interface Graph<T> {
 
     /**
-     * Add an attribute nodes to the entity graph.
+     * Add an attribute node to the entity graph.
      *
      * @param attributeName  name of the attribute
+     * @return the added attribute node
      * @throws IllegalArgumentException if the attribute is not an
      *         attribute of this entity.
      * @throws IllegalStateException if the EntityGraph has been
      *         statically defined
      *
      * @since 3.2
      */
-    void addAttributeNode(String attributeName);
+    <Y> AttributeNode<Y> addAttributeNode(String attributeName);
 
     /**
      * Add an attribute node to the entity graph.
      *
      * @param attribute  attribute
+     * @return the added attribute node
+     * @throws IllegalStateException if the EntityGraph has been
+     *         statically defined
+     *
+     * @since 3.2
+     */
+    <Y> AttributeNode<Y> addAttributeNode(Attribute<? super T, Y> attribute);
+
+    /**
+     * Get an existing attribute node for the attribute with the given
+     * name.
+     *
+     * @param attributeName  name of the attribute
+     * @return the attribute node
+     * @throws IllegalArgumentException if the attribute is not an
+     *         attribute of this entity.
+     * @throws java.util.NoSuchElementException if there is no existing
+     *         node for the attribute
+     *
+     * @since 3.2
+     */
+    <Y> AttributeNode<Y> getAttributeNode(String attributeName);
+
+    /**
+     * Get an existing attribute node for the given attribute.
+     *
+     * @param attribute  attribute
+     * @return the attribute node
+     * @throws java.util.NoSuchElementException if there is no existing
+     *         node for the attribute
+     *
+     * @since 3.2
+     */
+    <Y> AttributeNode<Y> getAttributeNode(Attribute<? super T, Y> attribute);
+
+    /**
+     * Get an existing attribute node for the attribute with the given
+     * name, or add an attribute node if there is no existing node.
+     *
+     * @param attributeName  name of the attribute
+     * @return the attribute node
+     * @throws IllegalArgumentException if the attribute is not an
+     *         attribute of this entity.
+     * @throws IllegalStateException if the EntityGraph has been
+     *         statically defined
+     *
+     * @since 3.2
+     */
+    <Y> AttributeNode<Y> getOrAddAttributeNode(String attributeName);
```

Comment:
The `Graph` has methods for adding and removing nodes, one method allowing iteration through existing nodes but no method for checking presence of the node. What if there are methods like `hasAttributeNode(String)` instead of `getOrAdd...`?

Reply:
`boolean` variant is better (and safer)

Reply:
Servido!

Reply:
Gracias

Reply:
@sebersole requested them.

I guess I see them as useful because `addAttributeNode()` would throw if there's an existing node (and you might not know that) .... except, _actually we don't say that explicitly_!

I wonder, @sebersole, if it would be enough to just give `addAttributeNode()` the semantics currently assigned to `getOrAddAttributeNode()`, instead of adding a separate operation.

Reply:
> I wonder, @sebersole, if it would be enough to just give addAttributeNode() the semantics currently assigned to getOrAddAttributeNode(), instead of adding a separate operation.

I think that's a good idea to combat method explosion.

Reply:
OK then, let's take that approach.